ledger/src/shred/merkle.rs

```diff
@@ -990,22 +1120,31 @@ pub(super) fn make_shreds_from_data(
 // shreds and attaches signature.
 fn make_erasure_batch(
     keypair: &Keypair,
-    shreds: Vec<ShredData>,
+    mut shreds: Vec<ShredData>,
+    // The Merkle root of the previous erasure batch if chained.
+    chained_merkle_root: Option<Hash>,
     next_code_index: u32,
     is_last_in_slot: bool,
     reed_solomon_cache: &ReedSolomonCache,
-) -> Result<Vec<Shred>, Error> {
+) -> Result<(/*merkle root:*/ Hash, Vec<Shred>), Error> {
     let num_data_shreds = shreds.len();
+    let chained = chained_merkle_root.is_some();
     let erasure_batch_size = shredder::get_erasure_batch_size(num_data_shreds, is_last_in_slot);
     let num_coding_shreds = erasure_batch_size - num_data_shreds;
     let proof_size = get_proof_size(erasure_batch_size);
     debug_assert!(shreds
         .iter()
-        .all(|shred| shred.common_header.shred_variant == ShredVariant::MerkleData(proof_size)));
+        .all(|shred| shred.common_header.shred_variant
+            == ShredVariant::MerkleData(proof_size, chained)));
     let mut common_header = match shreds.first() {
-        None => return Ok(Vec::default()),
+        None => return Err(Error::from(TooFewShards)),
         Some(shred) => shred.common_header,
     };
+    if let Some(hash) = chained_merkle_root {
+        for shred in &mut shreds {
+            shred.set_chained_merkle_root(&hash)?;
+        }
+    }
```

Comment:
Does it make sense to modify this closer to where the rest of the data shred is being modified here: https://github.com/solana-labs/solana/blob/9f44a29e8246543961721edbca5dedcf28807ec4/ledger/src/shred/merkle.rs#L1202

Reply:
We need to set the chained merkle root here because that is part of the payload which is hashed to generate merkle tree. i.e. we can't obtain:

     merkle tree => merkle root => signature

before setting the chained merkle root here.

Reply:
ah yeah that makes sense